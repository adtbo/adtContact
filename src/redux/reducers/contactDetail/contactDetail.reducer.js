import { createAction } from 'redux-actions';
import typeToReducer from 'type-to-reducer';

import { getContact, editContact, createContact } from '../../../services/contact/contact.service';

// Constants

const SET_CONTACT = 'DETAIL/SET_CONTACT';
const GET_CONTACT = 'DETAIL/GET_CONTACT';
const EDIT_CONTACT = 'DETAIL/EDIT_CONTACT';
const CREATE_CONTACT = 'DETAIL/CREATE_CONTACT';
const RESET_CONTACT = 'DETAIL/RESET_CONTACT';

export const actionTypes = {
    SET_CONTACT,
    GET_CONTACT,
    EDIT_CONTACT,
    CREATE_CONTACT,
    RESET_CONTACT
};

// Actions

export const setContactId = createAction(SET_CONTACT, contactId => Promise.resolve(contactId));
export const fetchContact = createAction(GET_CONTACT, getContact);
export const updateContact = createAction(EDIT_CONTACT, editContact);
export const createNewContact = createAction(CREATE_CONTACT, createContact);
export const resetContact = createAction(RESET_CONTACT);

export const actions = {
    setContactId,
    fetchContact,
    updateContact,
    createNewContact,
    resetContact,
};

// Reducer

// Initial State

export const initialState = {
    contactId: '',
    contact: {},
    error: {
        status: false,
        message: '',
    },
}

const setContactHandler = (state, action) => {
    return {
        ...state,
        contactId: action.payload
    }
}

const getAndUpdateContactHandler = (state, action) => {
    return {
        ...state,
        contact: action.payload.data
    }
};

const createContactHandler = (state, action) => {
    return {
        ...state,
    }
};

const errorHandler = (state, action) => {
    return {
        ...state,
        error:{
            status: true,
            message: action.payload.message
        }
    }
};

const resetContactHandler = (state) => {
    return {
        ...state,
        ...initialState
    }
}

// type to reducers
export default typeToReducer({
    [SET_CONTACT]:{
        FULFILLED: setContactHandler
    },
    [GET_CONTACT]: {
        FULFILLED: getAndUpdateContactHandler,
        REJECTED: errorHandler
    },
    [EDIT_CONTACT]:{
        FULFILLED: getAndUpdateContactHandler,
        REJECTED: errorHandler
    },
    [CREATE_CONTACT]:{
        FULFILLED: createContactHandler,
        REJECTED: errorHandler
    },
    [RESET_CONTACT]: resetContactHandler
}, initialState);