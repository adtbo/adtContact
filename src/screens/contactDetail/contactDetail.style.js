import { Platform, StyleSheet } from 'react-native';

export default StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
    },
    avaCont:{
        flex: 4,
        alignItems: 'center',
        justifyContent: 'center',
        borderBottomLeftRadius: 30,
        borderBottomRightRadius: 30,
        backgroundColor : '#E88C30'
    },
    labelStyle:{
        fontSize : 14,
        marginBottom : -6,
    },
    errorStyle:{
        fontSize : 10,
        color: 'red',
    },
    detailCont:{
        flex:3,
        alignItems: 'center',
        margin : 20,
        //borderWidth : 1,
    }
})