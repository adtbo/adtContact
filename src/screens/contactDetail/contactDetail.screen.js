import React, { useEffect, useState } from 'react'
import { Text, View, Image, ScrollView } from 'react-native'
import styles from './contactDetail.style'
import { useForm, Controller } from 'react-hook-form';

import { connect } from 'react-redux'
import { actions } from '../../redux/reducers/contactDetail/contactDetail.reducer'
import { actions as listActions } from '../../redux/reducers/contactList/contactList.reducer'
import { Avatar, Input, Button, Accessory } from 'react-native-elements'
import Spinner from 'react-native-loading-spinner-overlay'

export const ContactDetailScreen = props => {

  const {
    contactId,
    contact,
    error,
    getContact,
    createContact,
    updateContact,
    resetContact,
    defaultValues,
    navigation,
    getContacts,
    setSearch,
  } = props;

  const { control, handleSubmit, errors, setValue, getValues} = useForm({
    reValidateMode: 'onChange',
    defaultValues: {...defaultValues}
  })

  useEffect(() => {

    async function fetchData() {
      setIsLoading(true);
      if(!contactId){
        resetContact();
        setEditable(true);
      }
      else{
        await getContact(contactId);
      }
      setIsLoading(false);
    }

    fetchData();
  },[])

  useEffect(()=>{
    if(contactId){
      setValue("firstName", contact.firstName)
      setValue("lastName", contact.lastName)
      setValue("age", contact.age ? contact.age.toString() : null)
      setValue("photo", contact.photo)
    }
  },[contact]);

  const [ editable, setEditable ] = useState(false);
  const [ isLoading, setIsLoading] = useState(false);

  const onSubmit = async (data) => {
    if(contact.photo){
      data.photo = contact.photo;
    } else {
      data.photo = defaultValues.photo;
    }

    try {
      if(contactId){
        await updateContact(data, contactId);
      } else {
        await createContact(data);
      }
      await getContacts();
      await setSearch('');
      navigation.navigate('home')
    } catch (error){
      console.log(error.message)
    }
  }

  const urlRegex = new RegExp('^(https?:\\/\\/)?'+ // protocol
  '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
  '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
  '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*'+ // port and path
  '(\\?[;&a-z\\d%_.~+=-]*)?'+ // query string
  '(\\#[-a-z\\d_]*)?$','i'); // fragment locator

  return (
    <View style={styles.container}>
      <Spinner visible={isLoading}/>
      <View style={styles.avaCont}>
        <Avatar
          icon={{name: 'user', color:'#E88C30', type: 'font-awesome'}}
          rounded
          size={200}
          source= {urlRegex.test(contact.photo) ? {uri: contact.photo} : null}
          overlayContainerStyle= { {backgroundColor: 'white'}}
        >
          {editable && <Accessory
            name={'pencil'}
            type={'font-awesome'}
            size={40}
          />}
        </Avatar>
      </View>
      <View style={styles.detailCont}>
        <ScrollView style={{width : '80%'}}>
        <Controller
          as ={ Input }
          control = { control }
          id = "firstName"
          name = "firstName"
          onChange ={ args => args[0].nativeEvent.text}
          disabled = {!editable}
          errorMessage = {errors.firstName ? errors.firstName.message : ''}
          errorStyle = {styles.errorStyle}
          value = {contact.firstName}
          label = 'First Name'
          labelStyle = {styles.labelStyle}
          rules = 
            {{
              required: {value: true, message: "first name can't be blank" },
              validate: (value) => value.trim() !== '' || "first name can't be blank",
              pattern: {value: /^[a-zA-Z0-9]*$/, message: "must only contain alphanumeric characters"},
              minLength: {value: 3, message: "must be at least 3 characters"}
            }}
        />
        <Controller
          as ={ Input }
          control = { control }
          id = "lastName"
          name = "lastName"
          onChange ={ args => args[0].nativeEvent.text}
          disabled = {!editable}
          errorMessage = {errors.lastName ? errors.lastName.message : ''}
          errorStyle = {styles.errorStyle}
          value = {contact.lastName}
          label = 'Last Name'
          labelStyle = {styles.labelStyle}
          rules = 
            {{
              required: {value: true, message: "last name can't be blank" },
              validate: (value) => value.trim() !== '' || "last name can't be blank",
              pattern: {value: /^[a-zA-Z0-9]*$/, message: "must only contain alphanumeric characters"},
              minLength: {value: 3, message: "must be at least 3 characters"}
            }}
        />
        <Controller
          as ={ Input }
          control = { control }
          id = "age"
          name = "age"
          onChange ={ args => args[0].nativeEvent.text}
          disabled = {!editable}
          errorMessage = {errors.age ? errors.age.message : ''}
          errorStyle = {styles.errorStyle}
          value = {contact.age}
          label = 'Age'
          labelStyle = {styles.labelStyle}
          keyboardType = 'numeric'
          rules = 
            {{
              required: {value: true, message: "age can't be blank" },
              validate: (value) => value.trim() !== '' || "age can't be blank",
              pattern: {value: /^[0-9]*$/, message: "must only contain number"},
              min : {value: 1, message: "age must be more than 0"}
            }}
        />
        <Button
          title = { editable ? "Save" : "Edit" }
          buttonStyle = { editable ? {backgroundColor :'#109121', borderRadius: 100} : {backgroundColor :'#E88C30', borderRadius: 100} }
          onPress = { editable ? handleSubmit(onSubmit) : () => setEditable(true) }
        />
        </ScrollView>
      </View>
    </View>
  )
}

const mapStateToProps = state => ({
  contact: state.contactDetail.contact,
  error: state.contactDetail.error,
  contactId: state.contactDetail.contactId,
});

const mapDispatchToProps = {
  getContact: actions.fetchContact,
  updateContact: actions.updateContact,
  createContact: actions.createNewContact,
  resetContact: actions.resetContact,
  getContacts: listActions.fetchContactList,
  setSearch: listActions.setSearch,
};

ContactDetailScreen.defaultProps ={
  defaultValues:{
    firstName:'',
    lastName:'',
    age:'',
    photo:'https://www.digitaltveurope.com/files/2019/12/Baby-Yoda-Soup.jpg',
  }
}

export default connect( mapStateToProps, mapDispatchToProps )(ContactDetailScreen);