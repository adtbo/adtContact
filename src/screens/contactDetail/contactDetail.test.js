import * as React from 'react';
import renderer from 'react-test-renderer';
import { ContactDetailScreen } from './contactDetail.screen';
import { Input, Button } from 'react-native-elements'

let props = {
    contactId: '',
    contact: {},
    error: {},
    getContact: jest.fn(() => Promise.resolve({})),
    createContact: jest.fn(() => Promise.resolve({})),
    updateContact: jest.fn(() => Promise.resolve({})),
    resetContact: jest.fn(),
    defaultValues: {},
    navigation: {
        navigate: jest.fn()
    },
    getContacts: jest.fn(() => Promise.resolve({})),
    setSearch: jest.fn()
};

describe('ContactDetail form Screen Unit test', () => {
  let tree;

  jest.useFakeTimers();
  
  jest.mock('react-hook-form');

  it(`renders correctly for new contact`, async () => {
    await renderer.act(() => {
      tree = renderer.create(<ContactDetailScreen  { ...props} />).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });


  it(`renders correctly for update contact`, async () => {
    let newProps = JSON.parse(JSON.stringify(props));
    newProps.contactId= '1234'
    await renderer.act(() => {
      tree = renderer.create(<ContactDetailScreen  { ...newProps} />).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  describe('check each component',() => {

    const setState =jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, setState]);

    it('submit button', async () =>{
      let newProps = JSON.parse(JSON.stringify(props));
      newProps.contactId= '1234'
      newProps.contact={
        firstName : 'John',
        lastName : 'Doe',
        Age : 28,
        photo : 'some photo link'
      }
      await renderer.act(() => {
        component = renderer.create(<ContactDetailScreen { ...newProps } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const submitBtn = root.findAllByType(Button)[0].props;
        await submitBtn.onPress();
        expect(setState).toHaveBeenCalled();
        // await submitBtn.onPress();
        // expect(newProps.updateContact).toHaveBeenCalled();
      })
    })

    it('submit button', async () =>{
      await renderer.act(() => {
        component = renderer.create(<ContactDetailScreen { ...props } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const submitBtn = root.findAllByType(Button)[0].props;
        await submitBtn.onPress();
        
        // expect(props.createContact).toHaveBeenCalled();
      })
    })
  
    it('firstName input', async () =>{
      await renderer.act(() => {
        component = renderer.create(<ContactDetailScreen { ...props } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const textInpt = root.findAllByType(Input)[0].props;
        // Controller
        await textInpt.onChange({
            nativeEvent: {
              text: 'John'
            }
          });
  
        //console.log(textInpt)
        
        //expect(props.uploadAsset).toHaveBeenCalled();
      })
    })

    it('lastName input', async () =>{
      await renderer.act(() => {
        component = renderer.create(<ContactDetailScreen { ...props } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const textInpt = root.findAllByType(Input)[1].props;
        // Controller
        await textInpt.onChange({
            nativeEvent: {
              text: 'Doe'
            }
          });
  
        //console.log(textInpt)
        
        //expect(props.uploadAsset).toHaveBeenCalled();
      })
    })

    it('age input', async () =>{
      await renderer.act(() => {
        component = renderer.create(<ContactDetailScreen { ...props } />);
      });
  
      await renderer.act(async () => {
        const root = component.root;
        const textInpt = root.findAllByType(Input)[2].props;
        // Controller
        await textInpt.onChange({
            nativeEvent: {
              text: '28'
            }
          });
  
        //console.log(textInpt)
        
        //expect(props.uploadAsset).toHaveBeenCalled();
      })
    })
  });
});