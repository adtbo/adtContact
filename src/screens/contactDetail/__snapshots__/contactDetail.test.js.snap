// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactDetail form Screen Unit test renders correctly for new contact 1`] = `null`;

exports[`ContactDetail form Screen Unit test renders correctly for update contact 1`] = `null`;
